### PR TITLE
Update activitypub outline

### PR DESCRIPTION
**Pull request by Nutomic**
TODO:
- [ ] check that all the fields are updated after [refactoring](https://yerbamate.dev/LemmyNet/lemmy/pulls/112)
- [ ] check that mandatory fields are correct
- [x] write some introduction and description texts

### CODE REVIEW

**Comment by Nutomic** [comment]
Contains all activities now (and many TODOs). Still need to write some description texts and do cleanup.

File: docs/src/contributing_apub_api_outline.md [hunk 1]
```diff
@@ -1,380 +1,589 @@
...
 ```
-#### [Read](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-read)
-```
+
+| Field Name | Mandatory | Description |
+|---|---|---|
+| `actor` | yes | The user that is sending the follow request |
+| `object` | yes | The community to be followed |
+
+TODO: shouldn't the Follow also have `to`?
```
Comment (Nutomic) [inline docs/src/contributing_apub_api_outline.md:262]: Fixed in https://yerbamate.dev/LemmyNet/lemmy/commit/2ad60379e41b653cc4c841d024e2d9cdf7884f69

File: docs/src/contributing_apub_api_outline.md [hunk 2]
```diff
@@ -1,380 +1,592 @@
 # Activitypub API outline
 
-**This document is completely outdated and doesn't reflect the current state of Lemmy at all**
-
-- Start with the [reddit API](https://www.reddit.com/dev/api), and find [Activitypub vocab](https://www.w3.org/TR/activitystreams-vocabulary/) to match it.
-
-<!-- toc -->
```
Comment (dessalines) [inline docs/src/contributing_apub_api_outline.md:7]: Keep this toc and tocstop, then use markdown-toc:

```
npm i -g markdown-toc
markdown-toc -i contributing_apub_api_outline.md
```

This installs markdown-toc and runs it in place. Its good to have a TOC on a largo hierarchical doc like this.

File: docs/src/contributing_apub_api_outline.md [hunk 3]
```diff
@@ -1,380 +1,592 @@
...
-    + [Invite a moderator](#invite-a-moderator)
-    + [Accept Invitation](#accept-invitation)
-    + [Reject Invitation](#reject-invitation)
-
-<!-- tocstop -->
+This document is targeted at developers who are familiar with the ActivityPub and ActivityStreams protocols. It gives a detailed outline of the actors, objects and activities used by Lemmy.
+
+Before reading this, have a look at our [Federation Overview](contributing_federation_overview.md) to get an idea how Lemmy federation works on a high level.
+
+Lemmy does not yet follow the ActivityPub spec in all regards. For example, we don't set a valid context indicating our context fields. We also ignore fields like `inbox`, `outbox` or `endpoints` for remote actors, and assume that everything is Lemmy. For an overview of deviations, read [#698](https://github.com/LemmyNet/lemmy/issues/698). They will be fixed in the near future.
```
Comment (Nutomic) [inline docs/src/contributing_apub_api_outline.md:7]: Should also mention that Lemmy is currently completely unable to receive different formats, like array vs single value, or object id vs full object.

File: docs/src/contributing_apub_api_outline.md [hunk 4]
```diff
@@ -1,380 +1,592 @@
...
+| `content` | no | Text for the community sidebar, usually containing a description and rules |
+| `icon` | no | Icon, shown next to the community name |
+| `image` | no | Banner image, shown on top of the community page |
+| `inbox` | no | ActivityPub inbox URL |
+| `outbox` | no | ActivityPub outbox URL, only contains up to 20 latest posts, no comments, votes or other activities |
+| `followers` | no | Follower collection URL, only contains the number of followers, no references to individual followers |
+| `endpoints` | no | Contains URL of shared inbox |
+| `published` | no | Datetime when the community was first created |
+| `updated` | no | Datetime when the community was last changed |
+| `publicKey` | yes | The public key used to verify signatures from this actor |
```
Comment (dessalines) [inline docs/src/contributing_apub_api_outline.md:78]: These both look really good.

File: docs/src/contributing_apub_api_outline.md [hunk 5]
```diff
@@ -1,380 +1,592 @@
...
+        "owner": "https://enterprise.lemmy.ml/u/picard",
+        "publicKeyPem": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyH9iH83+idw/T4QpuRSY\n5YgQ/T5pJCNxvQWb6qcCu3gEVigfbreqZKJpOih4YT36wu4GjPfoIkbWJXcfcEzq\nMEQoYbPStuwnklpN2zj3lRIPfGLht9CAlENLWikTUoW5kZLyU6UQtOGdT2b1hDuK\nsUEn67In6qYx6pal8fUbO6X3O2BKzGeofnXgHCu7QNIuH4RPzkWsLhvwqEJYP0zG\nodao2j+qmhKFsI4oNOUCGkdJejO7q+9gdoNxAtNNKilIOwUFBYXeZJb+XGlzo0X+\n70jdJ/xQCPlPlItU4pD/0FwPLtuReoOpMzLi20oDsPXJBvn+/NJaxqDINuywcN5p\n4wIDAQAB\n-----END PUBLIC KEY-----\n"
+    }
 }
 ```
 
+| Field Name | Mandatory | Description |
+|---|---|---|
+| `name` | yes | Name of the actor |
+| `preferredUsername` | yes | Displayname |
```
Comment (dessalines) [inline docs/src/contributing_apub_api_outline.md:123]: This is an optional field. You have it right for community, where it's not-optional, but for users its optional.

File: docs/src/contributing_apub_api_outline.md [hunk 6]
```diff
@@ -1,380 +1,592 @@
...
+    "type": "Note",
+    "attributedTo": "https://enterprise.lemmy.ml/u/picard",
+    "to": "https://enterprise.lemmy.ml/c/main",
+    "content": "mmmk",
+    "inReplyTo": [
+        "https://enterprise.lemmy.ml/post/38",
+        "https://voyager.lemmy.ml/comment/73"
+    ],
+    "published": "2020-10-06T17:53:22.174836+00:00",
+    "updated": "2020-10-06T17:53:22.174836+00:00"
```
Comment (dessalines) [inline docs/src/contributing_apub_api_outline.md:192]: We don't need to change this here right now, but I think this is where you suggested we add a to / CC for the parent comment actor_id. Maybe add a TODO

File: docs/src/contributing_apub_api_outline.md [hunk 7]
```diff
@@ -1,380 +1,592 @@
...
 }
 ```
 
-### Communities
-#### [Create](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-create)
-```
+| Field Name | Mandatory | Description |
+|---|---|---|
+| `actor` | yes | The same community as in the `Follow` activity |
+| `to` | no | Inbox of the user which sent the `Follow` |
```
Comment (dessalines) [inline docs/src/contributing_apub_api_outline.md:284]: This is probably mandatory.

**Review by dessalines — CHANGES_REQUESTED** [review]

File: docs/src/contributing_apub_api_outline.md (hunk 3 above)
Comment (dessalines) [reply docs/src/contributing_apub_api_outline.md:7]: Sounds good.

File: docs/src/contributing_apub_api_outline.md (hunk 6 above)
Comment (Nutomic) [reply docs/src/contributing_apub_api_outline.md:192]: Right, that was in the comment create/update activity. I added it to #698.

File: docs/src/contributing_apub_api_outline.md (hunk 3 above)
Comment (Nutomic) [reply docs/src/contributing_apub_api_outline.md:7]: Done.

File: docs/src/contributing_apub_api_outline.md (hunk 2 above)
Comment (Nutomic) [reply docs/src/contributing_apub_api_outline.md:7]: Nice, done.

File: docs/src/contributing_apub_api_outline.md (hunk 5 above)
Comment (Nutomic) [reply docs/src/contributing_apub_api_outline.md:123]: Thanks, fixed.

File: docs/src/contributing_apub_api_outline.md (hunk 7 above)
Comment (Nutomic) [reply docs/src/contributing_apub_api_outline.md:284]: No its [not used](https://github.com/LemmyNet/lemmy/blob/main/lemmy_apub/src/inbox/user_inbox.rs#L90). But it should say "ID of the user", not inbox.

**Review by dessalines — APPROVED** [review]

**Pull request merged**